Fit parameters on history DAG

willdumm opened the pull request:
## Goal:
See #62 for motivation. This PR supersedes #62, because we can now calculate branching process parameters on the history DAG.

## Description of Changes:
* ambiguous sequences in trees output by dnapars are resolved in all possible (maximally parsimonious) ways
* resolved trees output by dnapars are used to create a history DAG, to which all allowed edges are added
* the DAG is trimmed so that it only expresses maximum parsimony trees
* an algorithm to eliminate edges between nodes with the same label (sequence) is run on the DAG
* the DAG is used to fit branching process parameters
* summary output is written to a log file for all trees in the DAG
* the history DAG, with all trees (not just max likelihood ones) is serialized and written to a file.
* the DAG is trimmed according to user-defined criteria
* One tree randomly sampled from each topology class in the trimmed DAG is rendered, etc. Mutability parsimony is a powerful discriminator, so when a mut model is provided there's likely to be only one tree in the trimmed DAG.


### DAG Trimming Details
By default, tree ordering is lexicographic, optimizing branching process likelihood, isotype parsimony, and mutability parsimony, in that order. Alternatively, a vector of weights for ranking priority may be provided, so that trees are instead ordered according to a linear combination of (transformed) likelihood, isotype parsimony, mutability parsimony, and number of alleles. `gctree infer` is endowed with new optional command line arguments for passing mutability and isotype information, as well as weights for ranking priority.

`gctree infer` now also takes either one or two positional arguments, allowing the user to pass the pickled history DAG object from a previous inference run, for example to quickly see the optimal tree when a different vector of ranking weights is provided. Another optional argument `--tree_only` was added to skip rewriting the (possibly very large) text file ranking all the trees in the DAG, making these quick exploratory trims slightly more practical.

### Other Details
* keeping track of when to call `_build_ll_genotype_cache` to avoid recursion errors was getting complicated, so I decided to let `_ll_genotype` manage its own preemptive caching. This is implemented by storing the max cached c and m for given parameters (p, q) in a class variable of `CollapsedTree`. On each call to `_ll_genotype`, if the passed c and m are greater than the maximum cached values for the passed parameters, the missing cache is built, avoiding calls to `_ll_genotype` that were made in earlier caching steps. Because of this, the new caching scheme should be at least as efficient as what we had before, and makes `_ll_genotype` easier to use. Disadvantages are that performance of `_ll_genotype` is less predictable (calling with uncached parameters can be unexpectedly slow), and caching is (like before) not targeted, making many more calls to `_ll_genotype` than necessary to avoid a recursion error.

## Tests
* passes existing tests, as well as many new tests in `tests/test_likelihoods.py`.

WSDeWitt commented on gctree/branching_processes.py:
```diff
@@ -987,3 +964,230 @@ def f(x):
     return result.x[0], result.x[1]
+
+
+def lltree(cm_counts, p: np.float64, q: np.float64) -> Tuple[np.float64, np.ndarray]:
```
Does this need to be a public function in the API that is shown in docs? If not, suggest renaming `_lltree`.

WSDeWitt commented on gctree/branching_processes.py:
```diff
@@ -987,3 +964,230 @@ def f(x):
+    return logf, grad_ll_genotype
+
+
+def llforest(
```
Does this need to be a public function in the API that is shown in docs? If not, suggest renaming `_llforest`.

WSDeWitt commented on gctree/branching_processes.py:
```diff
@@ -987,3 +964,230 @@ def f(x):
+        )
+
+
+def fit_branching_process(dag, verbose=True, marginal=True):
```
Does this need to be a public function in the API that is shown in docs? If not, suggest renaming `_fit_branching_process `.

willdumm replied:
I don't know, I had kind of thought that the inference pipeline should be reproducible using functions in the API. Like if you don't want to use the cli, all the pieces that it puts together are accessible and documented. I have no idea if that's a goal, or what sorts of things you want in the API, in general.

WSDeWitt replied:
Ok I agree with that. But note that this function isn't currently documented, because the `branching_processes` module isn't documented. The tree and forest classes in that module are promoted to package-level classes (see `gctree/__init__.py`), and those are documented. We'd need to add `branching_processes` in Modules section of `api.rst` (and probably a note in the Classes section explaining that these are promoted from the `branching_processes` module).

WSDeWitt replied:
Also note missing type hint on `dag` argument.

willdumm replied:
I think this is all taken care of, we'll see when I build the docs again.

WSDeWitt commented on gctree/branching_processes.py:
```diff
@@ -987,3 +964,230 @@ def f(x):
+    return (p, q)
+
+
+def clade_tree_to_ctree(
```
Does this need to be a public function in the API that is shown in docs? If not, suggest renaming `_clade_tree_to_ctree`

willdumm replied:
See reply to last comment

WSDeWitt replied:
(and missing type hint on `clade_tree` argument)

WSDeWitt commented on gctree/branching_processes.py:
```diff
@@ -987,3 +964,230 @@ def f(x):
+    return ctree
+
+
+def cmcounter_dagfuncs():
```
Does this need to be a public function in the API that is shown in docs? If not, suggest renaming `_ cmcounter_dagfuncs`

WSDeWitt commented on gctree/cli.py:
```diff
@@ -599,15 +542,16 @@ def get_parser():
+        ),
     )
     parser_infer.add_argument(
         "--bootstrap_phylipfile",
```
Is this argument still supported?

willdumm replied:
No. I will add a warning that it's not, as we already discussed 🙄. Thanks for the reminder

willdumm replied:
I've left it in, and it raises an informative error when passed. Maybe it should just be removed, though?

WSDeWitt replied:
We could just nuke all bootstrap stuff. This reminds me, the old scons pipeline should probably be nuked (as was forewarned in a previous release, since we have a CLI now)—it's probably very broken by now.

willdumm replied:
Sounds fun. I'll do that.

WSDeWitt commented on gctree/dag_filtering.py:
```diff
@@ -0,0 +1,507 @@
+r"""Functions for choosing optimal trees according to flexible criteria."""
```
Do we want this all to be a public module exposed in the API and docs for end users, or is much of this better kept under the hood?

willdumm replied:
I suppose the `filter_dag` function covers most of what someone might want to do. Having access to the filtering functions would make it easy to define custom criteria (e.g. some other branching process model likelihood, etc) and trim for that and an existing criterion jointly, or to trim according to the existing criteria (lexicographically) in a custom order. For example, you could do

```python
import historydag as hdag
import gctree.dag_filtering as df

with open('dagfile.p', 'rb') as fh:
    dag = hdag.deserialize(fh.read())

iso = df.isotype_dagfuncs(...)
mut = df.mutability_dagfuncs(...)
coolnewthing = hdag.utils.AddFuncDict(
    {
        # some functions...
    },
    names="coolnewthing"
)

dag.trim_optimal_weight(**(coolnewthing + iso + mut), optimal_func=min)
bp.CollapsedTree(dag.sampe().to_ete(...)).render(...)
```

This seems pretty slick to me, but that's not really justification for exposing it in the API, I guess. Maybe it's more in the scope of hDAG, and these filtering functions should be part of the hDAG api?

Do you think the current functionality of the `filter_dag` function is flexible enough?

WSDeWitt replied:
I agree that's slick, so I'll leave it to you to make the call on what should be public. I agree that more generic/abstract functionality could live in hdag, where you might want to reuse it.

willdumm replied:
Part of the reason I ended up moving all this to gctree was that having it in hdag required I import gctree to avoid duplicating (or moving) a bunch of fundamental gctree code. Then we're into circular dependency territory, so I think I'd better leave it here.

I think I'll make it public, and work on some better documentation. Now that I have historydag docs up, that should be easier.

WSDeWitt replied:
Sounds good. Can you point me at historydag docs? I could't find a link in the readme.

willdumm replied:
https://matsengrp.github.io/historydag/
Needs organization. Right now it's just built from docstrings with no context pages.

WSDeWitt replied:
Nice, thanks! You could link it in your README if you want it findable

WSDeWitt commented on gctree/mutation_model.py:
```diff
@@ -380,3 +371,22 @@ def simulate(
         return tree
+
+
+def sequence_disambiguations(sequence, _accum=""):
```
Do we want this public in the API?

WSDeWitt reviewed (commented):
Some initial high-level comments, mostly me sounding a conservative note about what is exposed in the API (but maybe me being hdag-naive at times). I have not dug into the copious technical details, but I see there are equally copious tests (yay!). Some tests are checking against code from the current implementation. This is great, but do we want to bring all of that along permanently? If we are convinced that this matches the previous likelihood code, we could consider losing those tests.

willdumm commented:
Thanks for the comments. I agree with almost everything. Sorry, I guess I should have thought through some of this more. It's good to have someone else look at it.

Regarding the tests that check against current code, I agree it's a bit inelegant and borderline redundant to drag that along. However, if there's further divergence in likelihood implementation in future versions, it would be nice to have that infrastructure in place, to immediately be sure that the likelihoods are still correct. I suspect that without those tests it would be easy to make breaking changes to likelihood computations, that don't fail any of the other tests.

I'd be interested in more elegant ideas for testing likelihood computations than copying old code. I think it's probably a good idea to have them be tested somehow, though.

willdumm commented:
One thing that's sort of a concern is memory usage when listing stats for all trees in the DAG. This is mostly because multisets of (c, m) pairs are accumulated for each tree, and that's expensive. I could use a less numerically stable likelihood computation that doesn't keep track of (c, m) pairs just for this part (but still use the stable one for the actual trimming step). The problem is that the resulting ranking of tree stats that's written to a file might not put the actual chosen best tree at the top, if calculated likelihoods which should be the same are slightly different.

If you have time to look into this, I think the most promising route to fix it would be to somehow improve the stability of `dag_filtering.ll_genotype_dagfuncs` (different numerical data types?), and/or perhaps it would be enough to round off a couple of decimal places of the tree log likelihoods before sorting, to eliminate the floating point arithmetic noise but hopefully keep all the real likelihood signal.

WSDeWitt commented on gctree/dag_filtering.py:
```diff
@@ -0,0 +1,507 @@
+
+    Args:
+        dag: A history DAG object to trim
+        priority_weights: A sequence of coefficients for prioritizing tree weights.
```
```suggestion
        priority_weights: A list or tuple of coefficients for prioritizing tree weights.
```

WSDeWitt commented on gctree/dag_filtering.py:
```diff
@@ -0,0 +1,507 @@
+        p, q: branching process parameters
+
+    Returns:
+        A ``historydag.utils.AddFuncDict`` object, which can be unpacked and used
```
Can we cross-reference to hdag docs here?

willdumm replied:
Yes. Those docs aren't up on GitHub pages yet, but I'll fix that.

WSDeWitt commented on gctree/dag_filtering.py:
```diff
@@ -0,0 +1,507 @@
+        branching process likelihood.
+    """
+
+    def edge_weight_ll_genotype(n1: hdag.HistoryDagNode, n2: hdag.HistoryDagNode):
+        """The _ll_genotype weight of the target node, unless it should be
+        collapsed, then 0.
+
+        Expects DAG to have abundances added with
+        :meth:`dag.add_abundances`.
+        """
+        if n2.is_leaf() and n2.label == n1.label:
+            return 0.0
+        else:
+            m = len(n2.clades)
+            # Check if this edge should be collapsed, and reduce mutant descendants
+            if frozenset({n2.label}) in n2.clades:
+                m -= 1
+            return bp.CollapsedTree._ll_genotype(n2.attr["abundance"], m, p, q)[0]
```
I'm having trouble understanding this function. Is `n2` the "focal node"? What is the `return 0.0` (i.e. probability 1) condition checking for? If we collapse an edge and decrement `m` don't we need to increment abundance by the label-matched child clade's abundance?

willdumm replied:
The assumptions are that the history DAG is collapsed, and that history DAG nodes have a supplementary label containing abundance information, so that each node with an observed sequence has the observed abundance associated with that sequence. When collapsing the history DAG, leaf-adjacent edges cannot be collapsed, so some trees in the DAG have uncollapsed tips.

We can think of the log likelihood of a CollapsedTree as the sum over all edges in that tree, of the log likelihood associated to the edge's target node's (c, m) pair. (we need to include the root node's likelihood to, of course)
This function is an implementation of that notion of edge weight. `n2` is the edge's target node, and the condition `return 0.0` corresponds to leaf adjacent edges which would be collapsed in a CollapsedTree, but aren't collapsed in the DAG. Because the parent node of such an edge has the same leaf sequence, it's already annotated by the same abundance. Therefore edges targeting that parent node will be assigned the correct abundance and `m` by the conditions in the else statement.

I'm not sure if that's very clear

WSDeWitt commented on gctree/branching_processes.py:
```diff
@@ -987,3 +964,253 @@ def f(x):
     return result.x[0], result.x[1]
+
+
+def _lltree(cm_counts, p: np.float64, q: np.float64) -> Tuple[np.float64, np.ndarray]:
```
Do we need this function? Can't its contents define the existing `CollapsedTree.ll` method?

willdumm replied:
We do need this, since computing likelihood on the hDAG bypasses CollapsedTree altogether, but does require the machinery in this function.

WSDeWitt commented on gctree/branching_processes.py:
```diff
@@ -987,3 +964,253 @@ def f(x):
+    return logf, grad_ll_genotype
+
+
+def _llforest(
```
Do we need this function? Can't its contents define the existing `CollapsedForest.ll` method?

WSDeWitt commented on gctree/branching_processes.py:
```diff
@@ -987,3 +964,253 @@ def f(x):
+        )
+
+
+def fit_branching_process(
```
I would think you could just replace the existing `CollapsedForest.mle` function with the contents of this function, instead of adding to the API

WSDeWitt reviewed (commented):
I added a few more single comments. Sorry if this is annoying, but I think there's too much exposure of the hdag implementation details to the user. There's also some inconsistency, since the `CollapsedForest` object has been retained in its original form (with its `mle` method), but new functional style content has been added to the API for using hdags. I note that in the CLI you no longer use `CollapsedForest.mle` for `gctree infer`, but it is still used in `gctree test`.

I think you could make this all more elegant (and more transparent to end users) by revising the `CollapsedForest` object to have an hdag attribute as it's implementation of a tree collection, and revise the `CollapsedForest.mle` method to do what your `fit_branching_process` currently does. If you did that, the CLI code won't need to be changed.

So my high-level recommendation is to incorporate the hdag as implementation changes of the existing API, to the extend possible. Thoughts?

willdumm commented:
I think that's a great idea. The object storing trees in the forest class will no longer be iterable, but I can add a method like `get_trees` which iterates through the stored CollapsedTrees. I'm not sure where the filtering code should go. Maybe the main `filter_dag` function can be a public method of `CollapsedForest`, and the rest of the filtering stuff can just be private functions somewhere?

It seems like it would still be nice to have a quick way of searching for the best tree with different criteria, without re-fitting. Is this something you'd rather not have in the CLI?

WSDeWitt commented:
That sounds good. It might be convenient to implement `__iter__` and `__next__` methods such that iterating on `CollapsedForest` object yields `CollapsedTree` objects from its `hdag` attribute?

I agree doing dag filtering via a public method is good. I suggest naming the public method `filter_trees` and avoid saying "DAG" in its docstring, just to make the method usable/understandable to end-users who don't know anything about hdags.

Can you clarify the last point about re-fitting? Do you mean accessing the `filter_trees` method via command line args to modify a previous `CollapsedForest` object provided as a pickle file, so you don't have to re-fit branching process parameters? I agree that could be useful if fitting branching process parameters takes much longer than dag filtering.

willdumm commented:
Yes, the last point was what you guessed. Is something like what's implemented now okay, where a single argument to gctree infer will be assumed to be a fit, CollapsedForest pickle file?

How do you want these changes to play with the CollapsedForest simulation stuff? I'm not sure it really makes sense to store simulated trees in a HistoryDag object. Maybe a CollapsedForest should have either a HistoryDag or a list of CollapsedTrees, and all the methods can manage the distinction under the hood?

WSDeWitt commented:
Yes the single argument looks good.

Good point about the simulation. The issue here is that we don't want the magic in the hdag that gives more trees than we used to seed it, we only want those seed trees. I wonder if this should be implemented on the hdag side? An hdag is a magic box that gives back more trees than you put in it, but here's a case (and there may be others) where we want it to work like a normal non-magic box, and give us back exactly the trees that we used to create it. Is there a straightforward way to retain that info in the hdag, and have an alternative iterator that generates only the seed trees?

willdumm commented:
No easy way to do that occurs to me, except for the hack of adding some dummy label field to force nodes that would otherwise be merged to be considered separate. That is, we'd create a DAG which is just shaped like a bunch of trees glued together at the root and at the leaves. Even this hack won't work though, because all the trees used to build an hDAG must have the same set of leaf labels (sequences, in this case). A forest of simulated trees wouldn't satisfy that property.

willdumm commented:
I think it will be relatively easy to manage with two separate data types, though. I will make `CollapsedForest.forest` private, so that nobody can access it thinking it's still a list of ctrees, and add another private variable `CollapsedForest._ctrees`.

willdumm commented on gctree/branching_processes.py:
```diff
@@ -62,7 +72,7 @@ def __init__(self, tree: ete3.TreeNode = None, allow_repeats: bool = False):
             observed_genotypes.add(self.tree.name)
```
This whole section can be drastically simplified I think, since we no longer need to accommodate bootstrap samples, and so maybe don't need to accumulate node names up the tree as sets. Is there any issue with removing this and just retaining one node name on edge collapse, no matter what?

Root-unifurcating trees added to the history DAG get an unobserved dummy leaf added below root (the history DAG doesn't allow unifurcation), so the logic in this section results in a CollapsedTree with a strange double-labeled root node after edge collapse. This can be fixed elsewhere (_clade_tree_to_ctree) if we don't want to change this section.

willdumm replied:
This is already resolved in `_clade_tree_to_ctree` so we can just ignore this whole idea for now if we want.

willdumm commented:
I'm occasionally getting the following error from check_grad, and it seems to be related to the latest underflow fix.
Just posting here so I can follow up later...

```
running gctree inference
gc90HK
Traceback (most recent call last):
  File "/home/wdumm/miniconda3/envs/gctree5/bin/gctree", line 33, in <module>
    sys.exit(load_entry_point('gctree', 'console_scripts', 'gctree')())
  File "/home/wdumm/gctree/gctree/gctree/cli.py", line 712, in main
    args.func(args)
  File "/home/wdumm/gctree/gctree/gctree/cli.py", line 172, in infer
    forest.mle(marginal=True)
  File "/home/wdumm/gctree/gctree/gctree/branching_processes.py", line 1035, in mle
    self.parameters = _mle_helper(self.ll, **kwargs)
  File "/home/wdumm/gctree/gctree/gctree/branching_processes.py", line 1418, in _mle_helper
    grad_check = sco.check_grad(lambda x: f(x)[0], lambda x: f(x)[1], x_0)
  File "/home/wdumm/miniconda3/envs/gctree5/lib/python3.9/site-packages/scipy/optimize/optimize.py", line 975, in check_grad
    approx_fprime(x0, func, step, *args))**2))
  File "/home/wdumm/miniconda3/envs/gctree5/lib/python3.9/site-packages/scipy/optimize/optimize.py", line 925, in approx_fprime
    return approx_derivative(f, xk, method='2-point', abs_step=epsilon,
  File "/home/wdumm/miniconda3/envs/gctree5/lib/python3.9/site-packages/scipy/optimize/_numdiff.py", line 486, in approx_derivative
    return _dense_difference(fun_wrapped, x0, f0, h,
  File "/home/wdumm/miniconda3/envs/gctree5/lib/python3.9/site-packages/scipy/optimize/_numdiff.py", line 557, in _dense_difference
    df = fun(x) - f0
  File "/home/wdumm/miniconda3/envs/gctree5/lib/python3.9/site-packages/scipy/optimize/_numdiff.py", line 437, in fun_wrapped
    f = np.atleast_1d(fun(x, *args, **kwargs))
  File "/home/wdumm/gctree/gctree/gctree/branching_processes.py", line 1418, in <lambda>
    grad_check = sco.check_grad(lambda x: f(x)[0], lambda x: f(x)[1], x_0)
  File "/home/wdumm/gctree/gctree/gctree/branching_processes.py", line 1416, in f
    return tuple(-y for y in ll(*x, **kwargs))
  File "/home/wdumm/gctree/gctree/gctree/branching_processes.py", line 1010, in ll
    scs.logsumexp((ls - ls[i_prime])[b != 0], b=b[b != 0])
  File "/home/wdumm/miniconda3/envs/gctree5/lib/python3.9/site-packages/scipy/special/_logsumexp.py", line 99, in logsumexp
    a_max = np.amax(a, axis=axis, keepdims=True)
  File "<__array_function__ internals>", line 180, in amax
  File "/home/wdumm/miniconda3/envs/gctree5/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 2791, in amax
    return _wrapreduction(a, np.maximum, 'max', axis, None, out,
  File "/home/wdumm/miniconda3/envs/gctree5/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 86, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
ValueError: zero-size array to reduction operation maximum which has no identity
```

willdumm commented on gctree/branching_processes.py:
```diff
@@ -920,42 +962,466 @@ def ll(
-                            ls - ls[i_prime], b=grad_ls[:, j] - grad_ls[i_prime, j]
-                        )
-                        - logsumexp(ls - ls[i_prime])
+                        scs.logsumexp((ls - ls[i_prime])[b != 0], b=b[b != 0])
```
This line is causing the error. When b consists entirely of zeros, then we're passing empty lists to logsumexp.  Not sure in what situation it's all zeros yet

willdumm replied:
fixed by `374cc06`. When the coefficient list `b` is all zeros, I was passing empty lists to logsumexp. Now I test for that situation and replace the offending calls with a hard-coded value

willdumm commented:
Here's a by-file summary of changes:

*deduplicate*
* fix bug in which sequences having different case aren't grouped correctly

*branching_processes*
* `CollapsedTree`
    * New attributes:
        * `_cm_list` became `_cm_counts`, a tuple of `(c, m)` counter items built by `_build_cm_counts` method
        * `_m_max` and `_c_max` became `_max_ll_cache` (class variable) a dictionary keyed by `(p,q)` tuples, with values `(c_max, m_max)` tuples, used in `_ll_genotype` method to compute-ahead and avoid recursion errors. (no more `_build_ll_genotype_cache`)
* `CollapsedForest`
    * New attributes:
        * `_forest` (A history DAG)
        * `_ctrees` (A list of CollapsedTrees if `_forest` is `None`). Wrapper `_requires_dag` asserts that some class methods aren't implemented for a CollapsedForest with tree data stored in `_ctrees`.
        * `_validation_stats` (A dictionary of data for checking CollapsedTrees extracted from `_forest`)
        * `_cm_countlist` (A tuple of `CollapsedTree._cm_list` counter items). This either gets populated when `_ctrees` is populated, or by `_forest` data when `ll` is first called. Both pathways independently manage adding root pseudocounts, if required.
    * New/changed methods and functions
        * `__init__` now takes a list of ete trees, makes a history DAG.  Alternatively, you can pass no tree data, and simulate a forest as before.
        * `filter_trees` handles filtering that was done directly in cli inference function before. Returns filtered forest and information about the trees in that forest
        * Implemented `__iter__` to make iterating through trees easy, regardless of whether they're stored as ctrees or in an hDAG. Also `__getstate__`, to exclude caches from pickling.
        * "dagfuncs" functions define the keyword arguments for counting weights in the DAG

*cli*
* reorganized and changed some arguments as discussed...

*isotyping*
* mostly added functions for computing isotype parsimony in DAG

*mutation_model*
* added mutability distance and DAG weight functions

*phylip_parse*
* minor changes...

WSDeWitt reviewed (approved):
Absolutely bonkers-level awesomeness detected—I hereby approve. :shipit:

willdumm commented:
closes #61, closes #70.
